src/cuda.cpp

```diff
@@ -108,46 +109,78 @@ bool stream_memory_operations_supported() {
 }
 
 FastEvent::FastEvent() {
-  if (!stream_memory_operations_supported()) {
-    throw_al_exception("FastEvent requires stream memory operations");
+  if (stream_memory_operations_supported()) {
+    sync_event = get_pinned_memory<int32_t>(1);
+    AL_CHECK_CUDA_DRV(cuMemHostGetDevicePointer(
+                        &sync_event_dev_ptr, sync_event, 0));
+  }
+  else {
+    plain_event = get_cuda_event();
   }
-  sync_event = get_pinned_memory<int32_t>(1);
-  AL_CHECK_CUDA_DRV(cuMemHostGetDevicePointer(
-                      &sync_event_dev_ptr, sync_event, 0));
 }
 
 FastEvent::~FastEvent() {
-  release_pinned_memory(sync_event);
+  if (stream_memory_operations_supported()) {
+    release_pinned_memory(sync_event);
+  }
+  else {
+    release_cuda_event(plain_event);
+  }
 }
 
 void FastEvent::record(cudaStream_t stream) {
-  // We cannot use std::atomic because we need the actual address of the memory.
-  __atomic_store_n(sync_event, 0, __ATOMIC_SEQ_CST);
-  AL_CHECK_CUDA_DRV(cuStreamWriteValue32(
-                      stream, sync_event_dev_ptr, 1,
-                      CU_STREAM_WRITE_VALUE_DEFAULT));
+  if (stream_memory_operations_supported()) {
+    // We cannot use std::atomic because we need the actual address of the memory.
+    __atomic_store_n(sync_event, 0, __ATOMIC_SEQ_CST);
+    AL_CHECK_CUDA_DRV(cuStreamWriteValue32(
+                        stream, sync_event_dev_ptr, 1,
+                        CU_STREAM_WRITE_VALUE_DEFAULT));
+  }
+  else {
+    AL_CHECK_CUDA(cudaEventRecord(plain_event, stream));
+  }
 }
 
 bool FastEvent::query() {
-  return __atomic_load_n(sync_event, __ATOMIC_SEQ_CST);
+  if (stream_memory_operations_supported())
+    return __atomic_load_n(sync_event, __ATOMIC_SEQ_CST);
+  else {
+    cudaError_t r = cudaEventQuery(plain_event);
+    if (r == cudaSuccess)
+      return true;
+    else if (r != cudaErrorNotReady)
+      throw_al_exception("FastEvent::query: cudaEventQuery error");
+    else
+      return false;
+  }
 }
 
-GPUWait::GPUWait() {
-  wait_sync = get_pinned_memory<int32_t>(1);
+GPUWait::GPUWait()
+  : wait_sync(get_pinned_memory<int32_t>(1)),
+    wait_sync_dev_ptr_no_stream_mem_ops(nullptr),
+    wait_sync_dev_ptr(0U)
+{
   // An atomic here may be overkill.
   // Can't use std::atomic because we need the actual address.
   __atomic_store_n(wait_sync, 0, __ATOMIC_SEQ_CST);
-  AL_CHECK_CUDA_DRV(cuMemHostGetDevicePointer(
-                      &wait_sync_dev_ptr, wait_sync, 0));
+
+  if (stream_memory_operations_supported())
+    AL_CHECK_CUDA_DRV(cuMemHostGetDevicePointer(
+                        &wait_sync_dev_ptr, wait_sync, 0));
+  else
+    AL_CHECK_CUDA(cudaHostGetDevicePointer(
```

Comment:
Why does it use cudaHostGetDevicePointer here and cuMemHostGetDevicePointer when stream memory operations are supported?

Reply:
The runtime and driver APIs actually represent device pointers slightly differently. The runtime API just treats them as pointers, but the driver API uses the `CUdeviceptr` type.

Reply:
OK, thanks.